### PR TITLE
Pespin/python3 support

### PR DESCRIPTION
**Pull request by pespin**

### CODE REVIEW

File: smpplib/command.py [hunk 1]
```diff
@@ -123,7 +124,7 @@ def generate_params(self):
         if hasattr(self, 'prep') and callable(self.prep):
             self.prep()
 
-        body = ''
+        body = b''
```
Comment (podshumok) [inline smpplib/command.py]: I think. it is better to define some byte-string constants and use them latter, like
in consts.py:
```
EMPTY_STRING=six.b('')
NULL_STRING=six.b('\0')
```

**Comment by podshumok** [comment]
I have second thoughts about using six.
At least we should add it as a requirement in setup.py
I think we should target python2.6 for now and be ready that some users don't have six or option to install it.
As we use very little functions from six, we can copy-paste them or, behave like that:
- try import needed functions from six
- if it fails, check python version
- if it is greater than 3, raise
- if no six and we are in python2 define `u`,`b`,`int2char` and `iteritems` as python2-only manually

**Comment by pespin** [comment]
I started looking at implementing the few bits we need, but I found out six module is not that big or complex, it's basically this file: https://bitbucket.org/gutworth/six/src/92e1c746e1abfbdb94705b821bc93766083efc54/six.py?at=default&fileviewer=file-view-default

Which means should not be an issue to install it on any system and depend on it. It will be better maintained than our one-time-copy-then-forget-about-it. Also, this way we get for free having to copy back more compatibility code later on if we need them when implementing new features or refactoring/redoing some code in the future.

So I propose the following for next version of the PR:
- We depend on six module (I check how to depend on it in setup.py)
- I define EMPTY_STRING / NULL_STRING as you commented
- I rebase it on top of current master

Waiting for you ACK to proceed on this.

File: setup.py [hunk 2]
```diff
@@ -23,6 +20,5 @@
         'Topic :: Communications :: Telephony',
         'Intended Audience :: Telecommunications Industry',
         'License :: OSI Approved',
-        ],
-      **extra
+        ]
```
Comment (podshumok) [inline setup.py]: please, add a coma

File: smpplib/command.py [hunk 3]
```diff
@@ -32,6 +33,8 @@
 
 logger = logging.getLogger('smpplib.command')
 
+EMPTY_STRING=six.b('')
+NULL_STRING=six.b('\0')
```
Comment (podshumok) [inline smpplib/command.py]: I think, It is better to put these to `consts.py`

File: smpplib/gsm.py [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
 
 # from http://stackoverflow.com/questions/2452861/python-library-for-converting-plain-text-ascii-into-gsm-7-bit-character-set
-gsm = (u"@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ\x1bÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>"
-       u"?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ`¿abcdefghijklmnopqrstuvwxyzäöñüà")
-ext = (u"````````````````````^```````````````````{}`````\\````````````[~]`"
-       u"|````````````````````````````````````€``````````````````````````")
+gsm = (six.u("@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ\x1bÆæßÉ !\"#¤%&'()*+,-./0123456789:;<=>"
+       "?¡ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÑÜ`¿abcdefghijklmnopqrstuvwxyzäöñüà"))
+ext = (six.u("````````````````````^```````````````````{}`````\\````````````[~]`"
+       "|````````````````````````````````````€``````````````````````````"))
```
Comment (podshumok) [inline smpplib/gsm.py]: Please fix the indentation

File: smpplib/gsm.py [hunk 5]
```diff
@@ -40,7 +41,7 @@ def make_parts(text):
         encoding = consts.SMPP_ENCODING_DEFAULT
         need_split = len(text) > consts.SEVENBIT_SIZE
         partsize = consts.SEVENBIT_MP_SIZE
-        encode = lambda s: s
+        encode = lambda s: six.b(s)
```
Comment (podshumok) [inline smpplib/gsm.py]: It is better to just say `encode = six.b`

File: smpplib/gsm.py [hunk 6]
```diff
@@ -60,9 +61,9 @@ def make_parts(text):
         ipart = 1
         uid = random.randint(0, 255)
         for start in starts:
-            parts.append(''.join(('\x05\x00\x03', chr(uid),
-                                  chr(len(starts)), chr(ipart),
-                                  encode(text[start:start + partsize]))))
+            parts.append( b''.join((b'\x05\x00\x03', six.int2byte(uid),
+                                  six.int2byte(len(starts)), six.int2byte(ipart),
+                                  encode(text[start:start + partsize]))) )
```
Comment (podshumok) [inline smpplib/gsm.py]: Please fix indentation

**Comment by podshumok** [comment]
So if it works for you in both python2 and python3, I'm merging

**Comment by pespin** [comment]
I am successfully sending sms with this using osmo-gsm-tester which uses python3 (code in osmo-gsm-tester which uses it is still not published, I have it in a local branch).

To test both python 2 and python 3, I am using the following test script and running it with python2 and python3. I am running an osmo-nitb which has an SMSC/SMPP interface, but with no MS connected to it, so I get an error "11 Invalid DestinationAddress" or similar, which is expected but other than that everything is fine, it connects + binds correctly and the SMSC parses the sms request correctly.

```
#!/usr/bin/python
# -*- coding: utf-8 -*-
import logging
import sys
import six

import smpplib.gsm
import smpplib.client
import smpplib.consts

# This is a small script using https://github.com/podshumok/python-smpplib
# to generate some SMS load on OsmoNITB via its SMPP interface

# if you want to know what's happening
logging.basicConfig(level='DEBUG')

def send_message(dest, string):
    parts, encoding_flag, msg_type_flag = smpplib.gsm.make_parts(string)

    #print 'Sending SMS "%s" to %s' % (string, dest)
    for part in parts:
        pdu = client.send_message(
            source_addr_ton=smpplib.consts.SMPP_TON_INTL,
            source_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
            source_addr='3802',
            dest_addr_ton=smpplib.consts.SMPP_TON_INTL,
            dest_addr_npi=smpplib.consts.SMPP_NPI_ISDN,
            destination_addr=dest,
            short_message=part,
            data_coding=encoding_flag,
            #esm_class=msg_type_flag,
            esm_class=smpplib.consts.SMPP_MSGMODE_FORWARD,
            registered_delivery=False,
    )
    print(pdu.sequence)

client = smpplib.client.Client('10.9.1.167', 2775)

# Print when obtain message_id
client.set_message_sent_handler(
    lambda pdu: sys.stdout.write('sent {} {}\n'.format(pdu.sequence, pdu.message_id)))
client.set_message_received_handler(
    lambda pdu: sys.stdout.write('delivered {}\n'.format(pdu.receipted_message_id)))

client.connect()
client.bind_transceiver(system_id='test', password='test')

destinations = ('3802')

for dest in destinations:
    send_message(dest, six.u('Mahlzeit'))
    send_message(dest, six.u('Руски'))
    send_message(dest, six.u('MahlzeitMahlzeitMahlzeitMahlzeitРуский из питона 3, кодировочку зажгиMahlzeitMahlzeitMahlzeitMahlzeitMРуский из питона 3, кодировочку зажгиahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMaРуский из питона 3, кодировочку зажгиhlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeiРуский из питона 3, кодировочку зажгиtMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeitMahlzeit'))

client.listen()
```

**Comment by podshumok** [comment]
Ok, lets hope everything is fine.

Great job! Thank you!

**Pull request merged**